airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/StatePersistence.java

```diff
@@ -0,0 +1,323 @@
+        state.getStateType() != StateType.LEGACY;
+
+    // The only case where we allow a state migration is moving from LEGACY.
+    // We expect any other migration to go through an explicit reset.
+    if (!isMigration && previousState.isPresent() && previousState.get().getStateType() != state.getStateType()) {
+      throw new IllegalStateException("Unexpected type migration from '" + previousState.get().getStateType() + "' to '" + state.getStateType() +
+          "'. Migration of StateType need to go through an explicit reset.");
+    }
+
+    this.database.transaction(ctx -> {
+      if (isMigration) {
+        clearLegacyState(ctx, connectionId);
+      }
+      switch (state.getStateType()) {
+        case GLOBAL -> saveGlobalState(ctx, connectionId, state.getGlobal().getGlobal());
+        case STREAM -> saveStreamState(ctx, connectionId, state.getStateMessages());
+        case LEGACY -> saveLegacyState(ctx, connectionId, state.getLegacyState());
+      }
+      return null;
+    });
+  }
+
+  private void clearLegacyState(final DSLContext ctx, final UUID connectionId) {
+    writeStateToDb(ctx, connectionId, null, null, StateType.LEGACY, null);
+  }
+
+  private void saveGlobalState(final DSLContext ctx, final UUID connectionId, final AirbyteGlobalState globalState) {
+    writeStateToDb(ctx, connectionId, null, null, StateType.GLOBAL, globalState.getSharedState());
+    for (final AirbyteStreamState streamState : globalState.getStreamStates()) {
+      writeStateToDb(ctx,
+          connectionId,
+          streamState.getStreamDescriptor().getName(),
+          streamState.getStreamDescriptor().getNamespace(),
+          StateType.GLOBAL,
+          streamState.getStreamState());
+    }
+  }
+
+  private void saveStreamState(final DSLContext ctx, final UUID connectionId, final List<AirbyteStateMessage> stateMessages) {
+    for (final AirbyteStateMessage stateMessage : stateMessages) {
+      final AirbyteStreamState streamState = stateMessage.getStream();
+      writeStateToDb(ctx,
+          connectionId,
+          streamState.getStreamDescriptor().getName(),
+          streamState.getStreamDescriptor().getNamespace(),
+          StateType.STREAM,
+          streamState.getStreamState());
+    }
+  }
+
+  private void saveLegacyState(final DSLContext ctx, final UUID connectionId, final JsonNode state) {
+    writeStateToDb(ctx, connectionId, null, null, StateType.LEGACY, state);
+  }
+
+  /**
+   * Performs the actual SQL operation depending on the state
+   *
+   * If the state is null, it will delete the row, otherwise do an insert or update on conflict
+   */
+  void writeStateToDb(final DSLContext ctx,
+                      final UUID connectionId,
+                      final String streamName,
+                      final String namespace,
+                      final StateType stateType,
+                      final JsonNode state) {
+    if (state != null) {
+      boolean hasState = ctx.selectFrom(STATE)
```

Comment:
can you add `final`, please? where possible we try to add `final` in the codebase. Here's how to set it up in intellij for it to lint it and do it automatically in some cases.

https://docs.airbyte.com/contributing-to-airbyte/code-style